### PR TITLE
State always reports changed when files and dirs exist at the same level

### LINKED ISSUE #51903: [Fluorine] State file.directory recurse: When both file_mode and dir_mode are supplied, file_mode overrides dir_mode for the directory itself
### Description of Issue/Question
Hi, I have an issue with file.directory state, when both file_mode and dir_mode are supplied, file_mode overrides dir_mode for the directory itself

According to [Saltstack Document](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.file.html#salt.states.file.directory), file_mode for permission of file, dir_mode for permission of directories.

### Setup
```
[root@localhost salttest]# pwd
/tmp/salttest
[root@localhost salttest]# tree -up .
.
├── [drwxr-xr-x root    ]  dir_one
│   └── [drwxr-xr-x root    ]  dir_two
│       └── [-rw-r--r-- root    ]  testfile.txt
├── [-rw-r--r-- root    ]  file_perm.sls
└── [-rw-r--r-- root    ]  top.sls

2 directories, 3 files

[root@localhost salttest]# cat file_perm.sls
/tmp/salttest/dir_one:
  file.directory:
    - user: root
    - group: root
    - dir_mode: 755
    - file_mode: 644
    - recurse:
      - user
      - group
      - mode

/tmp/salttest/dir_one/dir_two:
  file.directory:
    - user: root
    - group: root
    - dir_mode: 755
    - makedirs: True

```

### Steps to Reproduce Issue
```
[root@localhost salttest]# salt-call --local --file-root=/tmp/salttest/ state.apply test=True file_perm
local:
----------
          ID: /tmp/salttest/dir_one
    Function: file.directory
      Result: None
     Comment: The following files will be changed:
              /tmp/salttest/dir_one: mode - 0644
     Started: 14:41:07.828673
    Duration: 6.19 ms
     Changes:
----------
          ID: /tmp/salttest/dir_one/dir_two
    Function: file.directory
      Result: True
     Comment: The directory /tmp/salttest/dir_one/dir_two is in the correct state
     Started: 14:41:07.835033
    Duration: 0.738 ms
     Changes:

Summary for local
------------
Succeeded: 2 (unchanged=1)
Failed:    0
------------
Total states run:     2
Total run time:   6.928 ms

```

### Versions Report
```
[root@localhost salttest]# salt-call --versions-report
Salt Version:
           Salt: 2019.2.0

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Apr 11 2018, 07:36:10)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.5.1804 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-862.3.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.5.1804 Core
```

### PR DESCRIPTION
**Pull request by danfoster**
Fixes #51903

### What does this PR do?

Bugfix: When using the `file.directory` state with `resurse=[mode]` and you have directories and files at the same level, the state always reports as changed, even if no changes need to be made

### What issues does this PR fix or reference?

#51903

### Previous Behavior

```
$ umask 007
$ for i in `seq 1 3`; do mkdir -p /tmp/test/d$i; touch /tmp/test/f$i; done        
ls -l /tmp/test
total 0
drwx------ 2 root root 6 Mar  7 13:16 d1
drwx------ 2 root root 6 Mar  7 13:16 d2
drwx------ 2 root root 6 Mar  7 13:16 d3
-rw------- 1 root root 0 Mar  7 13:16 f1
-rw------- 1 root root 0 Mar  7 13:16 f2
-rw------- 1 root root 0 Mar  7 13:16 f3
$ salt-call state.single file.directory name="/tmp/test" dir_mode="700" file_mode="600" recurse="[mode]" test=false
local:
----------
          ID: /tmp/test
    Function: file.directory
      Result: True
     Comment: Directory /tmp/test is in the correct state
     Started: 13:54:50.775917
    Duration: 8.027 ms
     Changes:
              ----------
              /tmp/test/d1:
                  ----------
                  mode:
                      0600
              /tmp/test/d2:
                  ----------
                  mode:
                      0600
              /tmp/test/d3:
                  ----------
                  mode:
                      0600

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   8.027 ms
```

### New Behavior

```
$ salt-call state.single file.directory name="/tmp/test" dir_mode="700" file_mode="600" recurse="[mode]" test=false
local:

Summary for local
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:   8.128 ms
$ chmod 644 /tmp/test/f2
$ chmod 755 /tmp/test/d2
$ salt-call state.single file.directory name="/tmp/test" dir_mode="700" file_mode="600" recurse="[mode]" test=false
local:
----------
          ID: /tmp/test
    Function: file.directory
      Result: True
     Comment: Directory /tmp/test updated
     Started: 14:16:34.005604
    Duration: 11.362 ms
     Changes:
              ----------
              mode:
                  0700

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  11.362 ms
$ ls -l /tmp/test
total 0
drwx------ 2 root root 6 Mar  7 13:16 d1
drwx------ 2 root root 6 Mar  7 13:16 d2
drwx------ 2 root root 6 Mar  7 13:16 d3
-rw------- 1 root root 0 Mar  7 13:16 f1
-rw------- 1 root root 0 Mar  7 13:16 f2
-rw------- 1 root root 0 Mar  7 13:16 f3
```

### Tests written?

No

### Commits signed with GPG?

Yes

### CODE REVIEW

**Comment by ssiuhk** [comment]
Even recurse is not set for mode it's going to change as recurse_set check is not done for file_mode
Check https://github.com/saltstack/salt/blob/develop/salt/states/file.py#L722
```
[root@localhost tmp]# umask 0077
[root@localhost tmp]# for i in `seq 1 3`; do mkdir -p /tmp/test/d$i; touch /tmp/test/f$i; done
[root@localhost tmp]# ls -l /tmp/test/
total 0
drwx------. 2 root root 6 Mar  7 14:53 d1
drwx------. 2 root root 6 Mar  7 14:53 d2
drwx------. 2 root root 6 Mar  7 14:53 d3
-rw-------. 1 root root 0 Mar  7 14:53 f1
-rw-------. 1 root root 0 Mar  7 14:53 f2
-rw-------. 1 root root 0 Mar  7 14:53 f3
[root@localhost tmp]# chmod 644 /tmp/test/f2
[root@localhost tmp]# chmod 755 /tmp/test/d2
[root@localhost tmp]# salt-call --local state.single file.directory name="/tmp/test" dir_mode="700" file_mode="600" recurse="[user]" test=true
local:
----------
          ID: /tmp/test
    Function: file.directory
      Result: None
     Comment: The following files will be changed:
              /tmp/test/f2: mode - 0600
     Started: 14:53:44.998954
    Duration: 7.602 ms
     Changes:
              ----------
              /tmp/test/f2:
                  ----------
                  mode:
                      0600

Summary for local
------------
Succeeded: 1 (unchanged=1, changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   7.602 ms
```

**Comment by danfoster** [comment]
Thanks for the feedback. I've updated the PR to address that.

```
$ chmod 644 /tmp//test/f2
$ chmod 755 /tmp/test/d2
$ salt-call state.single file.directory name="/tmp/test" dir_mode="700" file_mode="600" recurse="[user]" test=false
local:

Summary for local
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:   7.717 ms
$ ls -l /tmp/test
total 0
drwx------ 2 root root 6 Mar  7 13:16 d1
drwxr-xr-x 2 root root 6 Mar  7 13:16 d2
drwx------ 2 root root 6 Mar  7 13:16 d3
-rw------- 1 root root 0 Mar  7 13:16 f1
-rw-r--r-- 1 root root 0 Mar  7 13:16 f2
-rw------- 1 root root 0 Mar  7 13:16 f3
$ salt-call state.single file.directory name="/tmp/test" dir_mode="700" file_mode="600" recurse="[mode]" test=false
local:
----------
          ID: /tmp/test
    Function: file.directory
      Result: True
     Comment: Directory /tmp/test updated
     Started: 15:56:15.106836
    Duration: 12.897 ms
     Changes:
              ----------
              mode:
                  0700

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  12.897 ms
$ ls -l /tmp/test
total 0
drwx------ 2 root root 6 Mar  7 13:16 d1
drwx------ 2 root root 6 Mar  7 13:16 d2
drwx------ 2 root root 6 Mar  7 13:16 d3
-rw------- 1 root root 0 Mar  7 13:16 f1
-rw------- 1 root root 0 Mar  7 13:16 f2
-rw------- 1 root root 0 Mar  7 13:16 f3
```

File: salt/states/file.py [hunk 1]
```diff
@@ -722,18 +722,18 @@ def _check_directory(name,
...
             if check_dirs:
                 for name_ in dirs:
                     path = os.path.join(root, name_)
-                    fchange = _check_dir_meta(path, user, group, mode, follow_symlinks)
+                    fchange = _check_dir_meta(path, user, group, dir_mode, follow_symlinks)
                     if fchange:
                         changes[path] = fchange
     # Recurse skips root (we always do dirs, not root), so always check root:
-    fchange = _check_dir_meta(name, user, group, mode, follow_symlinks)
+    fchange = _check_dir_meta(name, user, group, dir_mode, follow_symlinks)
```
Comment (ssiuhk) [inline salt/states/file.py]: dir_mode = None if 'mode' not in recurse_set, then you'll lost check for root_dir

File: salt/states/file.py [hunk 2]
```diff
@@ -676,7 +676,7 @@ def _check_directory(name,
                      user=None,
                      group=None,
                      recurse=False,
-                     mode=None,
+                     dir_mode=None,
```
Comment (ssiuhk) [inline salt/states/file.py]: I am not entirely sure if no other parts call this function by kwargs, it could break things

Comment (danfoster) [reply salt/states/file.py]: It looks like calls to this is contained within `states/file.py` and the calls here all use positional arguments:
```
 $ grep -r _check_directory .
./states/file.py:def _check_directory(name,
./states/file.py:def _check_directory_win(name,
./states/file.py:        presult, pcomment, pchanges = _check_directory_win(
./states/file.py:        presult, pcomment, pchanges = _check_directory(
./states/file.py:                check_res, check_msg, ret['pchanges'] = _check_directory_win(dirname)
./states/file.py:                check_res, check_msg, ret['pchanges'] = _check_directory(dirname)
./states/file.py:                check_res, check_msg, ret['pchanges'] = _check_directory_win(dirname)
./states/file.py:                check_res, check_msg, ret['pchanges'] = _check_directory(dirname)
```

File: salt/states/file.py (hunk 1 above)
Comment (danfoster) [reply salt/states/file.py]: moved the root_dir check to before the recurse check

**Comment by Ch3LL** [comment]
just fyi @danfoster I migrated your PR from 2018.3.4 to 2018.3. The 2018.3.4 release is already out so it will need to go into the 2018.3 branch and will be included in the next point release.

**Review by Ch3LL — CHANGES_REQUESTED** [review]
can we get a regression test added here?

**Comment by danlsgiga** [comment]
are we getting this in 2018.3.5? It affects us... I see it is already in 2019.2.1.

**Comment by Ch3LL** [comment]
Closing since 2018.3 is no longer receiving releases.

**Pull request closed without merging**